src/rust/bitbox02-rust/src/hww/api/ethereum/signmsg.rs

```diff
@@ -0,0 +1,217 @@
+// Copyright 2020 Shift Crypto AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::pb;
+use super::Error;
+
+use crate::workflow::verify_message;
+
+use pb::eth_response::Response;
+
+use alloc::vec::Vec;
+use core::convert::TryInto;
+
+use sha3::digest::Digest;
+
+/// Process a sign message request.
+///
+/// The result contains a 65 byte signature. The first 64 bytes are the secp256k1 signature in
+/// compact format (R and S values), and the last byte is the recoverable id (recid).
+pub async fn process(request: &pb::EthSignMessageRequest) -> Result<Response, Error> {
+    if request.msg.len() > 9999 {
+        return Err(Error::InvalidInput);
+    }
+    let pub_request = pb::EthPubRequest {
+        output_type: pb::eth_pub_request::OutputType::Address as _,
+        keypath: request.keypath.clone(),
+        coin: request.coin,
+        display: true,
+        contract_address: Vec::new(),
+    };
+
+    // Verify address. We don't need the actual result, but we have to propagate validation or user
+    // abort errors.
+    let _ = super::pubrequest::process(&pub_request).await?;
+
+    verify_message::verify(&request.msg).await?;
+
+    let mut msg: Vec<u8> = Vec::new();
+    msg.extend(b"\x19Ethereum Signed Message:\n");
+    msg.extend(format!("{}", request.msg.len()).as_bytes());
+    msg.extend(&request.msg);
+
+    let sighash: [u8; 32] = sha3::Keccak256::digest(&msg).as_slice().try_into().unwrap();
+    let sign_result = bitbox02::keystore::secp256k1_sign(&request.keypath, &sighash)?;
+
+    let mut signature: Vec<u8> = sign_result.signature.to_vec();
+    signature.push(sign_result.recid);
+
+    Ok(Response::Sign(pb::EthSignResponse { signature }))
+}
+
+#[cfg(test)]
```

Comment:
The C version also had some more tests like a too long msg. Maybe it's not so relevant anymore but I think I would appreciate some tests verifying a correct error is returned if say `keystore::secp256k1_sign` fails.

Reply:
done, added tests for both.